proposals/csharp-8.0/ranges.md

```diff
@@ -0,0 +1,58 @@
+# Ranges
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+
+This feature is about delivering two new operators that allow constructing `System.Index` and `System.Range` objects, and using them to index/slice collections at runtime.
+
+## Detailed Design
+
+#### System.Index
+
+C# has no way of indexing a collection from the end, but rather most indexers use the "from start" notion, or do a "length - i" expression. We introduce a new Index expression that means "from the end". The feature will introduce a new unary prefix "hat" operator. Its single operand must be convertible to `System.Int32`. It will be lowered into the appropriate `System.Index` factory method call.
+
+```csharp
+var thirdItem = list[2];                // list[2]
+var lastItem = list[^1];                // list[Index.CreateFromEnd(1)]
+
+var multiDimensional = list[3, ^2]      // list[3, Index.CreateFromEnd(2)]
+```
+
+#### System.Range
+
+C# has no syntactical way to access "ranges" or "slices" of collections. Usually users are forced to implement complex structures to filter/operate on slices of memory, or resort to LINQ methods like `list.Skip(5).Take(2)`. With the addition of `System.Span<T>` and other similar types, it becomes more important to have this kind of operation supported on a deeper level in the language/runtime, and have the interface unified.
+
+The language will introduce a new range operator `x..y`. It is a binary infix operator that accepts two expressions. Either operands can be omitted (examples below), and they have to be convertible to `System.Index`. It will be lowered to the appropriate `System.Range` factory method call.
+
+```csharp
+var slice1 = list[2..^3];               // list[Range.Create(2, Index.CreateFromEnd(3))]
```

Comment:
From our discussion, I suspect the `Range.Create` factory method will be sufficient, in the end. The `..` would translate to `Range.Create(Index.CreateFromStart(0), Index.CreateFromEnd(0))`.
I'll let you take a note for follow-up after next round of API design. Thanks